streams/src/test/java/org/apache/kafka/streams/state/internals/ListValueStoreTest.java

```diff
@@ -0,0 +1,174 @@
+        listStore.put(0, "zero");
+        // should retain duplicates
+        listStore.put(0, "zero again");
+        listStore.put(1, "one");
+        listStore.put(2, "two");
+
+        final KeyValue<Integer, String> zero = KeyValue.pair(0, "zero");
+        final KeyValue<Integer, String> zeroAgain = KeyValue.pair(0, "zero again");
+        final KeyValue<Integer, String> one = KeyValue.pair(1, "one");
+        final KeyValue<Integer, String> two = KeyValue.pair(2, "two");
+
+        assertEquals(
+            asList(zero, zeroAgain, one, two),
+            toList(listStore.all())
+        );
+    }
+
+    @Test
+    public void shouldGetAllNonDeletedRecords() {
+        // Add some records
+        listStore.put(0, "zero");
+        listStore.put(1, "one");
+        listStore.put(1, "one again");
+        listStore.put(2, "two");
+        listStore.put(3, "three");
+        listStore.put(4, "four");
+
+        // Delete some records
+        listStore.put(1, null);
+        listStore.put(3, null);
+
+        // Only non-deleted records should appear in the all() iterator
+        final KeyValue<Integer, String> zero = KeyValue.pair(0, "zero");
+        final KeyValue<Integer, String> two = KeyValue.pair(2, "two");
+        final KeyValue<Integer, String> four = KeyValue.pair(4, "four");
+
+        assertEquals(
+            asList(zero, two, four),
+            toList(listStore.all())
+        );
+    }
+
+    @Test
+    public void shouldGetAllReturnTimestampOrderedRecords() {
+        // Add some records in different order
+        listStore.put(4, "four");
+        listStore.put(0, "zero");
+        listStore.put(2, "two1");
+        listStore.put(3, "three");
+        listStore.put(1, "one");
+
+        // Add duplicates
+        listStore.put(2, "two2");
+
+        // Only non-deleted records should appear in the all() iterator
+        final KeyValue<Integer, String> zero = KeyValue.pair(0, "zero");
+        final KeyValue<Integer, String> one = KeyValue.pair(1, "one");
+        final KeyValue<Integer, String> two1 = KeyValue.pair(2, "two1");
+        final KeyValue<Integer, String> two2 = KeyValue.pair(2, "two2");
+        final KeyValue<Integer, String> three = KeyValue.pair(3, "three");
+        final KeyValue<Integer, String> four = KeyValue.pair(4, "four");
+
+        assertEquals(
+            asList(zero, one, two1, two2, three, four),
+            toList(listStore.all())
+        );
+    }
+
+    @Test
+    public void shouldEarlyClosedIteratorStillGetAllRecords() {
+        listStore.put(0, "zero");
+        listStore.put(1, "one");
+
+        final KeyValue<Integer, String> zero = KeyValue.pair(0, "zero");
+        final KeyValue<Integer, String> one = KeyValue.pair(1, "one");
+
+        final KeyValueIterator<Integer, String> it = listStore.all();
+        assertEquals(zero, it.next());
+        it.close();
+
+        // A new all() iterator after a previous all() iterator was closed should return all elements.
```

Comment:
Why is this a requirement?

Should we instead on call `valueList.removeAll()` on `iter.close()`? -- In this test, if there would be 3 keys, the third key won't be returned either after `close()` to why allow to keep iteration on the current key ?

Reply:
Not sure if I understand here... in the caller processor, we do not rely on the range to automatically delete values while iterating, but still explicitly call `put(k, null)` to do the delete, right?